File: Blind 75/18 - Bit Manipulation/190-reverse-bits.py
'''
Reverse bits of a given 32 bits unsigned integer.

Note:

- Note that in some languages, such as Java, there is no unsigned integer type. In this case, both input and output will be given as a signed integer type. They should not affect your implementation, as the integer's internal binary representation is the same, whether it is signed or unsigned.
- In Java, the compiler represents the signed integers using 2's complement notation. Therefore, in Example 2 above, the input represents the signed integer -3 and the output represents the signed integer -1073741825.
'''
from time import time


class Solution:
    def reverseBits(self, n: int) -> int:
        n = bin(n)[2:]
        n = '0' * (32 - len(n)) + n
        return int(n[::-1], 2)

    def reference(self, n: int) -> int:
        res = 0
        for i in range(32):
            bit = (n >> i) & 1
            res = res | (bit << (31 - i))
        return res

    def quantify(self, test_cases, runs=100000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.reverseBits(case))
                else:
                    self.reverseBits(case)
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        int('00000010100101000001111010011100', 2),
        int('11111111111111111111111111111101', 2),
    ]
    test.quantify(test_cases)

File: Blind 75/18 - Bit Manipulation/JavaScript/190-reverse-bits.js
const { performance } = require("perf_hooks");

class Solution {
  method(n) {
    let res = 0;
    let count = 32;

    while (count--) {
      res *= 2;
      res += n & 1;
      n = n >> 1;
    }
    return res;
  }

  reference(n, bit = 0) {
    for (let i = 0; i < 32; i++) {
      bit <<= 1; // Double * 2
      bit |= n & 1; // Flip
      n >>= 1; // Reduce * 0.5
    }

    return bit >>> 0;
  }

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  parseInt("00000010100101000001111010011100", 2),
  parseInt("11111111111111111111111111111101", 2),
];
test.quantify(testCases);
